ccp/src/cu/cu_prover.rs

```diff
@@ -161,22 +173,32 @@ impl CUProver {
         &'threads mut self,
         cache: CacheHandle,
         dataset: DatasetHandle,
+        cancel_token: CancellationToken,
     ) -> CUResult<()> {
         use futures::FutureExt;
 
-        let threads_number = self.threads.len() as u64;
         let dataset_size = dataset.items_count();
 
-        let closure = |thread_id: usize, thread: &'threads mut ProvingThreadAsync| {
-            let thread_id = thread_id as u64;
-
-            let start_item = (dataset_size * thread_id) / threads_number;
-            let next_start_item = (dataset_size * (thread_id + 1)) / threads_number;
-            let items_count = next_start_item - start_item;
-
-            thread
-                .initialize_dataset(cache.clone(), dataset.clone(), start_item, items_count)
-                .boxed()
+        let chunk_generator = PLMutex::new(ChunkGenerator::new(dataset_size, DATASET_CHUNK_SIZE));
```

Comment:
I think that you don't need a mutex here, just create separate chackers on equal non intersect intervals.

Reply:
Is it guaranteed that chunks of same size need same time for dataset generation?  If not, it will be more efficient to use a shared chunk generator.